### DISCUSSION TITLE
What should I do with submodules?

### DISCUSSION THREAD

**Discussion by HeavySnowJakarta** [community/community#69472] [New to GitHub] [answered]
Please forgive me for any unclearity but I'm trying my best to help you understand what happened. I'm trying to include submodules in my project, and the submodules imported belong to others, which were not forked. What's more, I added annotations as notes on the submodules, then I committed the changes to the repository of my project. When I tried to clone my repository from GitHub for another time, I found the directory of the submodules was empty. I tried to `git pull` but was told `Already up to date`. I tried to seek the changes I made from the GitHub website and the result is I can find my commits by clicking the submodule folder icon shown on the website, where there is a message `This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.`. Now I wonder if I am able to restore the commits on the submodules as I can find them on GitHub and whether I must fork a submodule if I want to make changes to it. For now, if I have to give up the changes, what should I do? Going back to the previous commit?

**Comment by moayad-star** [comment]
Regarding submodules: Don't bother with them, just leave the name or location of whatever module the user needs so that the size of the repo does not increase unnecessarily

**Comment by spenserblack** [comment, marked as answer]
It's best to think of submodules as separate repositories with their own histories. The "main" or "host" repository doesn't track the submodule history. Instead, it points to the commit that submodules should be checked out at.

If you're going to modify submodules, you probably want to do one of these:
- fork them so you have a copy that you own
- contribute to the subodule's remote repository via PRs.

But, basically, you don't want your local submodule to be checked out at a commit that the submodule's remote repository doesn't have.

> When I tried to clone my repository from GitHub for another time, I found the directory of the submodules was empty.

```shell
git submodule update --init
```
